### PR TITLE
Information filter

### PR DESCRIPTION
**Pull request by jmbarr**
The information form of the Kalman filter predictor and updater classes. Could be extended to encompass other members of the Kalman (EKF, UKF, etc) family.

### CODE REVIEW

File: stonesoup/predictor/information.py [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+        Returns
+        -------
+        : :class:`~.StateVector`
+            The predicted state vector
+
+        """
+        prior_state_mean = prior.covar @ prior.state_vector
+        return self.transition_model.matrix(**kwargs) @ prior_state_mean
+
+    @lru_cache()
```
Comment (sdhiscocks) [inline stonesoup/predictor/information.py:125]: Should use `predict_lru_cache` from `._utils` here.

File: .idea/codeStyles/codeStyleConfig.xml [hunk 2]
```diff
@@ -0,0 +1,5 @@
+<component name="ProjectCodeStyleConfiguration">
+  <state>
+    <option name="PREFERRED_PROJECT_CODE_STYLE" value="Default" />
+  </state>
+</component>
```
Comment (sdhiscocks) [inline .idea/codeStyles/codeStyleConfig.xml:5]: This file should be removed.

File: stonesoup/predictor/information.py [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+            These are passed to :meth:`~.LinearGaussianTransitionModel.matrix`
+
+        Returns
+        -------
+        : :class:`numpy.ndarray`
+            The inverse of the transition matrix, :math:`F_k^{-1}`
+
+        """
+        if hasattr(self.transition_model, 'inverse_transition_matrix'):
+            inv_transition_matrix = self.transition_model.inverse_matrix(**kwargs)
```
Comment (sdhiscocks) [inline stonesoup/predictor/information.py:94]:
```suggestion
        if hasattr(self.transition_model, 'inverse_matrix'):
            inv_transition_matrix = self.transition_model.inverse_matrix(**kwargs)
```

File: stonesoup/predictor/information.py [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+
+    """
+    transition_model = Property(
+        LinearGaussianTransitionModel,
+        doc="The transition model to be used.")
+    control_model = Property(
+        LinearControlModel,
+        default=None,
+        doc="The control model to be used. Default `None` where the predictor "
+            "will create a zero-effect linear :class:`~.ControlModel`.")
```
Comment (sdhiscocks) [inline stonesoup/predictor/information.py:77]:
```suggestion
    transition_model: LinearGaussianTransitionModel = Property(
        doc="The transition model to be used.")
    control_model: LinearControlModel = Property(
        default=None,
        doc="The control model to be used. Default `None` where the predictor "
            "will create a zero-effect linear :class:`~.ControlModel`.")
```

File: stonesoup/updater/information.py [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+    the existence of an :meth:`inverse_covar()` property. If absent, the :meth:`covar()` is
+    inverted.
+
+    """
+    measurement_model = Property(
+        LinearGaussian, default=None,
+        doc="A linear Gaussian measurement model. This need not be defined if "
+            "a measurement model is provided in the measurement. If no model "
+            "specified on construction, or in the measurement, then error "
+            "will be thrown.")
```
Comment (sdhiscocks) [inline stonesoup/updater/information.py:41]:
```suggestion
    measurement_model: LinearGaussian = Property(
        default=None,
        doc="A linear Gaussian measurement model. This need not be defined if "
            "a measurement model is provided in the measurement. If no model "
            "specified on construction, or in the measurement, then error "
            "will be thrown.")
```

File: stonesoup/types/state.py [hunk 6]
```diff
@@ -218,6 +218,18 @@ def covar(self):
...
+class InformationState(State):
+    r"""Information State Type
+
+    The information state class carries the :attr:`state_vector`,
+    :math:`\mathbf{y}_k = Y_k \mathbf{x}_k` and the precision or information matrix
+    :math:`Y_k = P_k^{-1}`, where :math:`\mathbf{x}_k` and :math:`P_k` are the mean and
+    covariance, respectively, of a Gaussian state.
+
+    """
+    precision = Property(PrecisionMatrix, doc='precision matrix of state.')
```
Comment (sdhiscocks) [inline stonesoup/types/state.py:230]:
```suggestion
    precision: PrecisionMatrix = Property(doc='precision matrix of state.')
```

File: stonesoup/updater/information.py [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+        \mathbf{y}_{k|k} = \mathbf{y}_{k|k-1} + H^{T}_k R^{-1}_k \mathbf{z}_{k}
+
+    where :math:`\mathbf{y}_{k|k-1}` is the predicted information state and :math:`Y_{k|k-1}` the
+    predicted information matrix which form the :class:`~.InformationStatePrediction` object. The
+    measurement matrix :math:`H_k` and measurement covariance :math:`R_k` are those in the Kalman
+    filter (see tutorial 1). An :class:`~.InformationStateUpdate` object is returned.
+
+    Note
+    ----
+    Analogously with the :class:`~.InformationStatePredictor`, the measurement model is queried for
```
Comment (sdhiscocks) [inline stonesoup/updater/information.py:31]:
```suggestion
    Analogously with the :class:`~.InformationKalmanPredictor`, the measurement model is queried for
```

File: stonesoup/updater/information.py [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+        pred_info_mean = hypothesis.prediction.state_vector
+        hh = measurement_model.matrix()
+        invr = self._inverse_measurement_covar(measurement_model)
+
+        posterior_precision = hypothesis.prediction.precision + hh.T @ invr @ hh
+        posterior_information_mean = pred_info_mean + hh.T @ invr @ \
+            hypothesis.measurement.state_vector
+
+        if force_symmetric_covariance:
+            posterior_precision = (posterior_precision + posterior_precision.T)/2
```
Comment (sdhiscocks) [inline stonesoup/updater/information.py:144]:
```suggestion
        if self.force_symmetric_covariance:
            posterior_precision = (posterior_precision + posterior_precision.T)/2
```

File: stonesoup/updater/information.py [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+
+        Parameters
+        ----------
+        hypothesis : :class:`~.SingleHypothesis`
+            the prediction-measurement association hypothesis. This hypothesis
+            carries a predicted information state.
+        force_symmetric_covariance : :obj:`bool`, optional
+            A flag to force the output precision to be symmetric by way
+            of a simple geometric combination of the matrix and transpose.
+            Default is `False`
```
Comment (sdhiscocks) [inline stonesoup/updater/information.py:120]:
```suggestion
    def update(self, hypothesis, **kwargs):
        r"""The Information filter update (corrector) method. Given a hypothesised association
        between a predicted information state and an actual measurement, calculate the posterior
        information state.

        Parameters
        ----------
        hypothesis : :class:`~.SingleHypothesis`
            the prediction-measurement association hypothesis. This hypothesis
            carries a predicted information state.
```

**Review by sdhiscocks — CHANGES_REQUESTED** [review]
A f ew changes required to match up with existing Kalman filter and type hints, but otherwise looks good :+1:

File: stonesoup/updater/information.py [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+            The predicted state in information form :math:`\mathbf{y}_{k|k-1}`
+        measurement_model : :class:`~.MeasurementModel`
+            The measurement model. If omitted, the model in the updater object
+            is used
+        **kwargs : various
+            These are passed to :meth:`~.MeasurementModel.matrix()`
+
+        Returns
+        -------
+        : :class:`GaussianMeasurementPrediction`
```
Comment (nperree-dstl) [inline stonesoup/updater/information.py:85]:
```suggestion
        : :class:`~.GaussianMeasurementPrediction`
```

**Review by nperree-dstl — APPROVED** [review]
Found one link that didn't work, otherwise it looks good and all makes sense!

**Review by sdhiscocks — APPROVED** [review]

**Pull request merged**